R/org_classify_description_texts.R

```diff
@@ -107,15 +109,66 @@ get_aws_url <- function(ein, year = 2019, form = NULL, move_global = TRUE) {
     
   }
   
-  out <- obj_id %>% 
-    select(ObjectId, SubmittedOn) %>%
-    mutate(tax_year = gsub("-.*", "", SubmittedOn)) %>%
-    arrange(desc(tax_year, SubmittedOn)) %>%
-    filter(tax_year == year) %>%
-    select(ObjectId)
+  df <- obj_id %>% 
+    select(ObjectId, TaxPeriod, SubmittedOn) 
+  
+  # One tax period 
+  
+  n_tp <- length(unique(df$TaxPeriod))
+  
+  if (n_tp >= 2) {message(glue("Multiple tax periods are found."))}
+  
+  if (n_tp == 1) {  
+  
+    # Filter tax period
+    df %>%
+      mutate(TaxPeriod_ch = as.character(TaxPeriod)) %>%
+      filter(str_detect(TaxPeriod_ch, as.character(tax_period))) # tax_period is a new argument
+  
+    # Select the latest submitted one 
+    out <- df %>%
+      dplyr::arrange(desc(SubmittedOn)) %>%
+      dplyr::slice(1)
+  
+    # Glue search parameter and the rest of the URL together
+    return(glue("http://s3.amazonaws.com/irs-form-990/{out$ObjectId}_public.xml"))
```

Comment:
Your comments are spot-on. I have thought about this problem, but haven't explained the way I think about it. 

It's correct that if `get_aws_url()` returns a dataframe, the following steps won't work. Here I would like users to determine which tax period they should use by inspecting the dataframe outcome. Put differently, multiple urls argument is an option for data exploration. 

However, if I'm making a counterpoint against myself, perhaps it's easier to sort these things out when users have the whole dataset in their hands. 

Honestly, I've avoided this problem because I didn't want to change all the rest of the steps ... but it seems there's a more straightforward solution (like you suggested by taking URL as an option.) Here I said this new feature is an "option (= argument)" because changing from taking EIN to a year to URL might also break some other things. Let me know if I misunderstood something in this context. 

For your other related comment, yes, if there are multiple unique tax periods, the function will return a dataframe of more than 1 row. But one thing unclear to me is if we insist on having one row, how can we select one row? 

Regarding the example you provided, I guess the best way to tackle that is group by tax period and run the function on the list of EINs, assuming that all these EINs are unique in each tax period. 

Having written this, I think that some of these issues stem from architectural concerns. We might want to separate the issues we need to better deal with the function opposed to database perspectives.